### PR TITLE
chore: drop Python 3.5 & 3.6 support

### PR DESCRIPTION
**Pull request by vytas7**
~~This is work in progress. I created a Draft PR in order to be able to verify progress with CI.~~

I had to stop short of some planned improvements in favour of getting the bulk of 3.5 & 3.6 removals out, I'll address these things in follow-up PRs.

### CODE REVIEW

File: falcon/constants.py [hunk 1]
```diff
@@ -8,18 +8,8 @@
...
-
-# TODO(vytas): Remove these hacks in 4.0 once we have dropped 3.5/3.6 support.
-# NOTE(vytas): Do not alter the microversion if already on 3.5
-#   (read: a hack to hit coverage on 3.5 as well).
-if _mock_py35 or _is_py35:
-    PYTHON_VERSION = (3, 5, PYTHON_VERSION[-1] if _is_py35 else 0)
-
-ASGI_SUPPORTED = PYTHON_VERSION >= (3, 6)
+
+ASGI_SUPPORTED = True
```
Comment (CaselIT) [inline falcon/constants.py:12]: I guess removing this would be a backward incompatible change. I guess it will not cause problems to keep it here

**Review by CaselIT — APPROVED** [review]

File: falcon/constants.py (hunk 1 above)
Comment (vytas7) [reply falcon/constants.py:12]: Yeah it wasn't underscored so I thought to play it safe.

**Comment by vytas7** [comment]
I'm still polishing various things... Need to clean up the tests and undo some foul import hacks.

File: falcon/util/__init__.py [hunk 2]
```diff
@@ -76,24 +76,18 @@
...
-
-else:
-    import json  # NOQA
+
+def __getattr__(name):
+    if name == 'json':
+        import warnings
+        import json  # NOQA
+
+        warnings.warn(
```
Comment (CaselIT) [inline falcon/util/__init__.py:85]: Maybe we could just remove this?

**Review by CaselIT — APPROVED** [review]

File: falcon/util/__init__.py (hunk 2 above)
Comment (vytas7) [reply falcon/util/__init__.py:85]: Good idea, but we need to add a breaking change newsfragment then telling what was removed. I can do this in a follow-up PR.

Comment (CaselIT) [reply falcon/util/__init__.py:85]: Fine by me

**Pull request merged**